src/main/java/org/prebid/server/bidder/huaweiads/CountryCodeResolver.java

```diff
@@ -0,0 +1,362 @@
+package org.prebid.server.bidder.huaweiads;
+
+import com.iab.openrtb.request.BidRequest;
+import com.iab.openrtb.request.Device;
+import com.iab.openrtb.request.Geo;
+import com.iab.openrtb.request.User;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+class CountryCodeResolver {
+
+    private static final Map<String, String> ALPHA3_TO_ALPHA2_COUNTRY_CODE_MAP = createAlpha3ToAlpha2CountryCodeMap();
+    private static final Map<Integer, String> MCC_TO_COUNTRY_CODE_MAP = createMccToAlpha2CountryCodeMap();
+
+    private CountryCodeResolver() {
+
+    }
+
+    //expects Alpha3 country code
+    public static Optional<String> resolve(BidRequest bidRequest) {
+        final Optional<String> countryOfDevice = Optional.ofNullable(bidRequest.getDevice())
+                .map(Device::getGeo)
+                .map(Geo::getCountry)
+                .filter(StringUtils::isNotBlank)
+                .filter(countryCode -> countryCode.length() >= 2)
+                .map(CountryCodeResolver::convertAlpha3ToAlpha2CountryCode);
+
+        final Optional<String> countryOfUser = Optional.ofNullable(bidRequest.getUser())
+                .map(User::getGeo)
+                .map(Geo::getCountry)
+                .filter(StringUtils::isNotBlank)
+                .filter(countryCode -> countryCode.length() >= 2)
+                .map(CountryCodeResolver::convertAlpha3ToAlpha2CountryCode);
+
+        final Optional<String> countryOfMcc = Optional.ofNullable(bidRequest.getDevice())
+                .map(Device::getMccmnc)
+                .filter(StringUtils::isNotBlank)
+                //mcc-mnc format
+                //todo: what if out of bound
+                .map(mccmnc -> mccmnc.split("-")[0])
+                .flatMap(CountryCodeResolver::countryFromMcc);
+
+        //todo: logic is slighty enhanced comparing to the PBS Go,
+        // if a country code of the device is invalid (has less than 3 symbols) I try to get the country from the user
+        // if the country of the user is invalid (has less than 3 symbols) I try to get the country from the MCC
+        // PBS Go returns default country code instead without trying to try to resolve code further
```

Comment:
This logic should do exactly the same as go version.

Reply:
From the perspective of logic it doesn't make sense to resolve the default country without trying to fetch a real country, I guess it's just missing in PBS Go